Firestore/core/src/model/base_path.h

```diff
@@ -174,6 +184,31 @@ class BasePath {
 
  private:
   SegmentsT segments_;
+
+  static util::ComparisonResult compareSegments(const std::string& lhs,
+                                                const std::string& rhs) {
+    bool isLhsNumeric = isNumericId(lhs);
+    bool isRhsNumeric = isNumericId(rhs);
+
+    if (isLhsNumeric && !isRhsNumeric) {
+      return util::ComparisonResult::Ascending;
+    } else if (!isLhsNumeric && isRhsNumeric) {
+      return util::ComparisonResult::Descending;
+    } else if (isLhsNumeric && isRhsNumeric) {
+      return util::Compare(extractNumericId(lhs), extractNumericId(rhs));
+    } else {
+      return util::Compare(lhs, rhs);
+    }
+  }
+
+  static bool isNumericId(const std::string& segment) {
+    return segment.substr(0, 4) == "__id" &&
+           segment.substr(segment.size() - 2) == "__";
+  }
+
+  static int64_t extractNumericId(const std::string& segment) {
+    return std::stol(segment.substr(4, segment.size() - 2));
```

Comment:
`std::stol` throws an `std::invalid_argument` exception if the argument is an empty string. So it reinforces that this function should *not* be called if the segment length is 6 or smaller.